File: parallax-ai/backend/app/schemas/simulation.py
from typing import Literal

from pydantic import BaseModel, Field


ConfidenceLevel = Literal[
    "low",
    "medium",
    "high",
]

ScenarioType = Literal[
    "best_case",
    "base_case",
    "worst_case",
]

EvidenceType = Literal[
    "known",
    "inferred",
    "assumed",
    "unknown",
]


class OutcomeDimension(BaseModel):
    name: str
    direction: Literal[
        "strong_positive",
        "positive",
        "neutral",
        "negative",
        "strong_negative",
        "uncertain",
    ]
    explanation: str


class FutureScenario(BaseModel):
    id: str
    option: str
    scenario_type: ScenarioType
    title: str

    narrative: str

    outcome_dimensions: list[OutcomeDimension]

    key_drivers: list[str]
    assumptions_required: list[str]
    warning_signals: list[str]

    reversibility: Literal[
        "low",
        "medium",
        "high",
    ]

    uncertainty: ConfidenceLevel

    robustness_score: float = Field(
        ge=0,
        le=100,
        description=(
            "Comparative PARALLAX scenario robustness score. "
            "Not a probability."
        ),
    )


class OptionSimulation(BaseModel):
    option: str
    futures: list[FutureScenario]


class SimulationResult(BaseModel):
    simulations: list[OptionSimulation]


class AssumptionItem(BaseModel):
    id: str
    statement: str
    category: EvidenceType

    source: str | None = None

    confidence: ConfidenceLevel

    impact: Literal[
        "low",
        "medium",
        "high",
        "critical",
    ]

    affected_options: list[str]

    validation_question: str | None = None


class AssumptionLedger(BaseModel):
    items: list[AssumptionItem]


class BlindSpot(BaseModel):
    id: str
    title: str
    explanation: str

    severity: Literal[
        "low",
        "medium",
        "high",
        "critical",
    ]

    affected_options: list[str]

    why_it_matters: str

    validation_action: str


class BlindSpotReport(BaseModel):
    blind_spots: list[BlindSpot]


class StressScenario(BaseModel):
    id: str
    name: str
    description: str

    severity: Literal[
        "moderate",
        "severe",
        "extreme",
    ]


class StressImpact(BaseModel):
    option: str
    scenario_id: str

    impact: Literal[
        "benefits",
        "minor_damage",
        "moderate_damage",
        "major_damage",
        "failure",
        "uncertain",
    ]

    explanation: str

    survives: bool


class StressTestResult(BaseModel):
    scenarios: list[StressScenario]
    impacts: list[StressImpact]

    most_resilient_option: str | None

    explanation: str


class CounterfactualCondition(BaseModel):
    variable: str
    required_change: str
    explanation: str


class CounterfactualResult(BaseModel):
    current_preference: str
    challenger: str

    conditions: list[CounterfactualCondition]

    plausible: bool

    summary: str


class InformationValueItem(BaseModel):
    question: str
    reason: str

    decision_impact: Literal[
        "low",
        "medium",
        "high",
        "critical",
    ]

    effort_to_obtain: Literal[
        "low",
        "medium",
        "high",
    ]

    affected_options: list[str]


class InformationValueResult(BaseModel):
    highest_value_question: str
    items: list[InformationValueItem]

File: parallax-ai/frontend/src/types/simulation.ts
export type ScenarioType = 
    | "best_case"
    | "base_case"
    | "worst_case";

export type Direction = 
    | "strong_positive"
    | "positive"
    | "neutral"
    | "negative"
    | "strong_negative"
    | "uncertain";

export interface OutcomeDimension {
    name: string;
    direction: Direction;
    explanation: string;
}

export interface FutureScenario {
    id: string;
    option: string;
    scenario_type: ScenarioType;
    title: string;
    narrative: string;
    outcome_dimensions: OutcomeDimension[];
    key_drivers: string[];
    assumptions_required: string[];
    warning_signals: string[];

    reversibility: "low" | "medium" | "high";
    uncertainty: "low" | "medium" | "high";
    robustness_score: number;
}

export interface OptionSimulation {
    option: string;
    futures: FutureScenario[];
}

export interface FutureSimulationPayload {
    simulation: {
        simulations: OptionSimulation[];
    };

    inference: {
        tier: string;
        model: string;
        latency_ms: number;

        usage: {
            prompt_tokens: number;
            completion_tokens: number;
            total_tokens: number;
        };
    };
}

/*Assumption, blind spots, and stress tests types*/
export type EvidenceCategory =
  | "known"
  | "inferred"
  | "assumed"
  | "unknown";

export interface AssumptionItem {
  id: string;
  statement: string;
  category: EvidenceCategory;
  source: string | null;
  confidence:
    | "low"
    | "medium"
    | "high";
  impact:
    | "low"
    | "medium"
    | "high"
    | "critical";
  affected_options: string[];
  validation_question: string | null;
}

export interface AssumptionLedgerPayload {
  ledger: {
    items: AssumptionItem[];
  };

  inference: {
    tier: string;
    model: string;
    latency_ms: number;

    usage: {
      prompt_tokens: number;
      completion_tokens: number;
      total_tokens: number;
    };
  };
}

export interface BlindSpot {
  id: string;
  title: string;
  explanation: string;
  severity:
    | "low"
    | "medium"
    | "high"
    | "critical";
  affected_options: string[];
  why_it_matters: string;
  validation_action: string;
}

export interface BlindSpotPayload {
  report: {
    blind_spots: BlindSpot[];
  };

  inference: {
    tier: string;
    model: string;
    latency_ms: number;

    usage: {
      prompt_tokens: number;
      completion_tokens: number;
      total_tokens: number;
    };
  };
}

export interface StressScenario {
  id: string;
  name: string;
  description: string;
  severity:
    | "moderate"
    | "severe"
    | "extreme";
}

export interface StressImpact {
  option: string;
  scenario_id: string;

  impact:
    | "benefits"
    | "minor_damage"
    | "moderate_damage"
    | "major_damage"
    | "failure"
    | "uncertain";

  explanation: string;

  survives: boolean;
}

export interface StressTestPayload {
  stress_test: {
    scenarios: StressScenario[];
    impacts: StressImpact[];
    most_resilient_option:
      | string
      | null;
    explanation: string;
  };

  inference: {
    tier: string;
    model: string;
    latency_ms: number;

    usage: {
      prompt_tokens: number;
      completion_tokens: number;
      total_tokens: number;
    };
  };
}

/** Counter-factual, Information, AI Jury */
export interface CounterfactualCondition {
  id: string;
  option: string;
  condition: string;
  explanation: string;

  likelihood:
    | "low"
    | "medium"
    | "high";

  impact:
    | "moderate"
    | "high"
    | "decisive";

  observable_signal: string;
}

export interface CounterfactualPayload {
  counterfactual: {
    conditions: CounterfactualCondition[];
    summary: string;
  };

  inference: {
    tier: string;
    model: string;
    latency_ms: number;

    usage: {
      prompt_tokens: number;
      completion_tokens: number;
      total_tokens: number;
    };
  };
}


export interface InformationItem {
  id: string;
  question: string;
  why_it_matters: string;

  information_value:
    | "low"
    | "medium"
    | "high"
    | "critical";

  effort:
    | "low"
    | "medium"
    | "high";

  affected_options: string[];
  suggested_action: string;
}

export interface InformationValuePayload {
  analysis: {
    items: InformationItem[];
    highest_value_question:
      | string
      | null;
    summary: string;
  };

  inference: {
    tier: string;
    model: string;
    latency_ms: number;

    usage: {
      prompt_tokens: number;
      completion_tokens: number;
      total_tokens: number;
    };
  };
}


export interface JuryVote {
  agent: string;
  preferred_option: string;
  confidence: number;
  rationale: string;
  strongest_argument: string;
  biggest_concern: string;
}

export interface JuryPayload {
  jury: {
    votes: JuryVote[];

    consensus_option:
      | string
      | null;

    agreement_score: number;

    disagreement_summary: string;

    minority_report:
      | string
      | null;
  };

  inference?: {
    tier: string;
    model: string;
    latency_ms: number;

    usage?: {
      prompt_tokens: number;
      completion_tokens: number;
      total_tokens: number;
    };
  };
}